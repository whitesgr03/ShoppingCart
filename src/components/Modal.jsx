import { useState } from "react";
import { Link } from "react-router-dom";

import Icon from "@mdi/react";
import { mdiTrashCanOutline } from "@mdi/js";

const Modal = ({
	cartList,
	showModal,
	latestItem,
	onToggleModal,
	onEditItem,
	onRemoveItem,
	onSetLatestItem,
}) => {
	const [deleteItem, setDeleteItem] = useState(null);

	const handleCheckTarget = e =>
		(e.target.classList.contains("shadow") ||
			e.target.classList.contains("close")) &&
		handleCloseModal();

	const handleCloseModal = () => {
		onToggleModal(null);
		setDeleteItem(null);
		onSetLatestItem(null);
	};

	return (
		<div
			data-testid="shadow"
			className={`shadow ${showModal ? `${showModal} active` : ""}`}
			onPointerUp={handleCheckTarget}
		>
			{cartList && (
				<Cart
					cartList={cartList}
					onEditItem={onEditItem}
					onToggleModal={onToggleModal}
					onSetDeleteItem={setDeleteItem}
				/>
			)}
		</div>
	);
};

const Cart = ({ cartList, onEditItem, onToggleModal, onSetDeleteItem }) => {
	const handleRemove = product => {
		onToggleModal("showAlert");
		onSetDeleteItem(product);
	};

	const totalPrice = cartList
		.reduce((sum, product) => sum + product.price * product.quantity, 0)
		.toFixed(2);

	const createOptions = maxQuantity => {
		const quantity = maxQuantity > 10 ? maxQuantity : 10;

		return Array.from({ length: quantity })
			.fill([])
			.map((_, i) => (
				<option key={i} value={i + 1}>
					{i + 1}
				</option>
			));
	};

	const list = cartList.map(product => (
		<div className="item" key={product.id}>
			<div className="warp">
				<img src={product.url} alt={product.name} />
			</div>
			<div className="description">
				<div className="info">
					<div className="name">{product.name}</div>
					<div className="price">${product.price.toFixed(2)}</div>
				</div>

				<div className="subtotal">
					<label htmlFor={product.name}>
						Quantity
						<select
							data-testid="quantity"
							name="quantity"
							id={product.name}
							onChange={e => {
								onEditItem(product.id, +e.target.value);
							}}
							value={product.quantity}
						>
							{createOptions(product.quantity)}
						</select>
					</label>
					<p>
						Subtotal: $
						{(product.quantity * product.price).toFixed(2)}
					</p>
				</div>
			</div>
			<button
				className="removeBtn"
				data-testid="removeBtn"
				onPointerUp={() => handleRemove(product)}
			>
				<Icon path={mdiTrashCanOutline} />
			</button>
		</div>
	));

	return (
		<div className="cart">
			<button className="close"></button>
			<div className="title">Shopping cart</div>

			{cartList && cartList.length > 0 ? (
				<>
					<div className="items">{list}</div>
					<div className="checkout">
						<p className="price">
							Total: <b>${totalPrice}</b>
						</p>

						<Link className="checkoutBtn slide" to={"/checkout"}>
							CHECKOUT
						</Link>
					</div>
				</>
			) : (
				<div data-testid="empty" className="empty">
					Your cart is currently empty.
				</div>
			)}
		</div>
	);
};

const Alert = ({
	newItem,
	deleteItem,
	onToggleModal,
	onSetLatestItem,
	onRemoveItem,
	onSetDeleteItem,
}) => {
	const product = (newItem || deleteItem) ?? false;

	const handleShowCart = () => {
		onToggleModal("showCart");
		removeAlertProduct();
	};

	const handleRemoveItem = () => {
		onRemoveItem(product.id);
		handleShowCart();
	};

	const removeAlertProduct = () => {
		onSetLatestItem(null);
		onSetDeleteItem(null);
	};

	return (
		<div className="alert">
			{product && (
				<>
					<div className="title" data-testid="title">
						{newItem
							? "Add product to cart"
							: "Remove product from cart"}
					</div>
					<button className="close"></button>
					<div className="product">
						{
							<>
								<div className="image">
									<img src={product.url} alt={product.name} />
								</div>
								<div className="description">
									<h2 className="title">{product.name}</h2>
									<p className="quantity">
										Quantity: {product.quantity}
									</p>
									<p className="price">
										Price: ${product.price.toFixed(2)}
									</p>
									<p className="subtotal">
										Subtotal (without tax):{" "}
										{product.quantity}
										item(s) $
										{(
											product.quantity * product.price
										).toFixed(2)}
									</p>
								</div>
							</>
						}
					</div>
					<div className="buttons">
						{newItem ? (
							<>
								<button
									className="left slide"
									onPointerUp={handleShowCart}
								>
									VIEW CART
								</button>
								<button className="right slide close">
									CONTINUE SHOPPING
								</button>
							</>
						) : (
							<>
								<button
									className="left slide"
									onPointerUp={handleRemoveItem}
								>
									REMOVE
								</button>
								<button
									className="right slide"
									onPointerUp={handleShowCart}
								>
									CANCEL
								</button>
							</>
						)}
					</div>
				</>
			)}
		</div>
	);
};

export { Modal as default, Cart, Alert };
