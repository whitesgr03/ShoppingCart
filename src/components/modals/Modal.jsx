import "../../style/modals/modal.css";

import { useState } from "react";

import ModalCart from "./ModalCart";
import ModalProductAlert from "./ModalProductAlert";

import { useModal, useModalDispatch } from "../../App/RootContext";

const Modal = () => {
	const [isLoading, setIsLoading] = useState(false);

	const { modal } = useModal();

	const dispatch = useModalDispatch();

	const handleCloseModule = e => {
		const classList = e.target.classList;

		!isLoading &&
			(classList.contains("shadow") || classList.contains("close")) &&
			dispatch({
				type: "close",
			});
	};

	return (
		<div className="modal">
			<div
				data-testid="shadow"
				className={`shadow ${modal ? `${modal} active` : ""}`}
				onPointerUp={handleCloseModule}
			>
				<ModalCart isLoading={isLoading} onLoading={setIsLoading} />
				<ModalProductAlert
					isLoading={isLoading}
					onLoading={setIsLoading}
				/>
			</div>
		</div>
	);
};

export { Modal as default };
