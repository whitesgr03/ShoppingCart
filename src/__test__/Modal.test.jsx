import { screen, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { RouterProvider, createMemoryRouter } from "react-router-dom";

import Modal, { Cart, Alert } from "../components/Modal";

describe("Renders Modal Component", () => {
	it("Should return Modal DOM", () => {
		const routes = [
			{
				path: "/",
				element: <Modal />,
			},
		];

		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});

		render(<RouterProvider router={router} />);

		const actual = screen.getByTestId("shadow");

		expect(actual).toBeInTheDocument();
	});

	it("Should close Modal DOM with click event", async () => {
		const mockToggleModal = jest.fn();
		const mockSetLatestItem = jest.fn();

		const user = userEvent.setup();
		const routes = [
			{
				path: "/",
				element: (
					<Modal
						onToggleModal={mockToggleModal}
						onSetLatestItem={mockSetLatestItem}
					/>
				),
			},
		];

		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});

		render(<RouterProvider router={router} />);

		const shadow = screen.getByTestId("shadow");

		await user.pointer({ keys: "[MouseLeft]", target: shadow });

		expect(mockToggleModal).toBeCalledTimes(1);
		expect(mockSetLatestItem).toBeCalledTimes(1);
	});
});

describe("Renders Cart Component", () => {
	it("Should return empty class without data", () => {
		const mockCartList = [];

		render(<Cart cartList={mockCartList} />);

		const actual = screen.getByTestId("empty");

		expect(actual).toBeInTheDocument();
	});
	it("Should return cart list with data", () => {
		const mockCartList = [
			{
				id: 0,
				name: "fake",
				url: "../",
				price: 19.9,
				quantity: 1,
			},
		];

		const routes = [
			{
				path: "/",
				element: <Cart cartList={mockCartList} />,
			},
		];

		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});

		render(<RouterProvider router={router} />);

		expect(screen.getByRole("img")).toHaveAttribute("src", "../");

		expect(screen.getByTestId("quantity")).toHaveValue("1");

		expect(screen.getByText("fake")).toHaveClass("name");

		expect(screen.getByText("Total:")).toHaveClass("price");
	});
	it("Should show Remove Alert modal DOM with click event", async () => {
		const mockToggleModal = jest.fn();
		const mockSetDeleteItem = jest.fn();

		const user = userEvent.setup();
		const mockCartList = [
			{
				id: 0,
				name: "fake",
				url: "../",
				price: 19.9,
				quantity: 1,
			},
		];

		const routes = [
			{
				path: "/",
				element: (
					<Cart
						cartList={mockCartList}
						onToggleModal={mockToggleModal}
						onSetDeleteItem={mockSetDeleteItem}
					/>
				),
			},
		];

		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});

		render(<RouterProvider router={router} />);

		const button = screen.getByTestId("removeBtn");

		await user.pointer({ keys: "[MouseLeft]", target: button });

		expect(mockToggleModal).toBeCalledTimes(1);
		expect(mockSetDeleteItem).toBeCalledTimes(1);
	});
	it("Should change select value with change event", async () => {
		const mockEditItem = jest.fn();

		const user = userEvent.setup();
		const mockCartList = [
			{
				id: 0,
				name: "fake",
				url: "../",
				price: 19.9,
				quantity: 1,
			},
		];

		const routes = [
			{
				path: "/",
				element: (
					<Cart cartList={mockCartList} onEditItem={mockEditItem} />
				),
			},
		];

		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});

		render(<RouterProvider router={router} />);

		const select = screen.getByTestId("quantity");

		expect(select).toHaveValue("1");

		await user.selectOptions(select, "2");

		expect(mockEditItem).toBeCalledTimes(1);
	});
});

describe("Renders Alert Component", () => {
	it("Should return Alert DOM", () => {
		const { container } = render(<Alert />);

		expect(container).toMatchSnapshot();
	});

	it("Should return Alert DOM with title 'Add product to cart'", () => {
		const mockNewItem = {
			id: 0,
			name: "fake",
			url: "../",
			price: 19.9,
		};
		render(<Alert newItem={mockNewItem} />);

		const actual = screen.getByTestId("title");

		const expected = "Add product to cart";

		expect(actual).toHaveTextContent(expected);
	});
	it("Should return Alert DOM with title 'Remove product from cart'", () => {
		const mockDeleteItem = {
			id: 0,
			name: "fake",
			url: "../",
			price: 19.9,
		};
		render(<Alert deleteItem={mockDeleteItem} />);

		const actual = screen.getByTestId("title");

		const expected = "Remove product from cart";

		expect(actual).toHaveTextContent(expected);
	});

	it("Should show cart DOM with click event", async () => {
		const user = userEvent.setup();

		const mockToggleModal = jest.fn();
		const mockSetLatestItem = jest.fn();
		const mockSetDeleteItem = jest.fn();

		const mockNewItem = {
			id: 0,
			name: "fake",
			url: "../",
			price: 19.9,
		};

		render(
			<Alert
				newItem={mockNewItem}
				onToggleModal={mockToggleModal}
				onSetLatestItem={mockSetLatestItem}
				onSetDeleteItem={mockSetDeleteItem}
			/>
		);

		const button = screen.getByRole("button", { name: "VIEW CART" });

		await user.pointer({ keys: "[MouseLeft]", target: button });

		expect(mockToggleModal).toBeCalledTimes(1);
		expect(mockSetLatestItem).toBeCalledTimes(1);
		expect(mockSetDeleteItem).toBeCalledTimes(1);
	});

	it("Should remove product with click event", async () => {
		const user = userEvent.setup();

		const mockToggleModal = jest.fn();
		const mockSetLatestItem = jest.fn();
		const mockSetDeleteItem = jest.fn();
		const mockRemoveItem = jest.fn();

		const mockDeleteItem = {
			id: 0,
			name: "fake",
			url: "../",
			price: 19.9,
		};

		render(
			<Alert
				deleteItem={mockDeleteItem}
				onToggleModal={mockToggleModal}
				onSetLatestItem={mockSetLatestItem}
				onSetDeleteItem={mockSetDeleteItem}
				onRemoveItem={mockRemoveItem}
			/>
		);

		const button = screen.getByRole("button", { name: "REMOVE" });

		await user.pointer({ keys: "[MouseLeft]", target: button });

		expect(mockToggleModal).toBeCalledTimes(1);
		expect(mockSetLatestItem).toBeCalledTimes(1);
		expect(mockSetDeleteItem).toBeCalledTimes(1);
		expect(mockRemoveItem).toBeCalledTimes(1);
	});
});
