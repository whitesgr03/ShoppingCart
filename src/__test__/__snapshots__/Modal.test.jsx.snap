// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Alert Component Should return Alert DOM 1`] = `
<div>
  <div
    class="alert"
  />
</div>
`;
